static/js/imageBuilder.js

```diff
@@ -204,12 +205,42 @@
 
   function render() {
     renderSummary();
+    updateForm();
+    checkDisabled();
   }
 
-  function changeState(key, value) {
-    STATE[key] = value;
+  function changeState(key, value, action) {
```

Comment:
It seems to me you're writing a lot of code to turn the interface from a declarative method-based one into an obscure function that expects strings with very specific required values that are hard to guess.

How are these functions meaningfully different from the default options available on an array?

``` javascript
const STATE = {
  'board': [],
  'os': [],
  'snaps': []
};

# UPDATE
STATE.board = ["new", "list"];

# ADD
STATE.board.push("newvalue")
```

I can't quite tell what you're intending with "REMOVE" - are you expecting `value` to contain the *index* of an item to remove? That seems odd, but if so then `STATE.board.splice(3, 1)` seems potentially clearer than `changeState("board", 3, "REMOVE")`.

If you wanted to be able to do `STATE.board.remove("somevalue")`, which seems like a reasonable thing to want to do, then that could be achieved with:

``` javascript
class StateArray extends Array{
  remove(itemValue) {
    const index = array.indexOf(5);
    if (index > -1) {
      this.splice(index, 1);
    }
  }
}

const STATE = {
  'board': StateArray(),
  'os': StateArray(),
  'snaps': StateArray()
};
```

But I also have a few other questions:

- Why is `STATE` upper-case? I know it's technically a `const` but you're hardly treating it as generally immutable, which is normally what upper case would indicate...
- Do you need state at all? You could simply do e.g. `boardState = StateArray()`
- Why do your initial values for `STATE` include 2 strings and 1 array? Surely you'd want all your states to be of the same basic type...

Reply:
After trying to implement these changes for a while today I ran into a number of issues:
- There is no native reassignment function on arrays therefore I created an update function but that basically called remove and then push which seemed awkward.
- The board and os variables are only ever a single value and so I would have to get the values with `state.board[0]` throughout the code which I really don't like.

If you feel strongly @nottrobin I am happy to give it another whirl but this branch has been in for a while. Let me know what you think?

Reply:
I played around with `watch`, which is deprecated, and then `__defineSetter__` etc., but the best I could come up with was the more straightforward approach, something like this:

``` javascript
class StateArray extends Array {
  push(item) {
    const result = super.push(item);
    render();
    return result;
  }
  pop() {
    const item = super.pop();
    render();
    return item;
  }
  remove(item) {
    const index = this.indexOf(item);
    if (index > -1) {
      const item = this.splice(index, 1);
    }
    render();
    return item;
  }
}

class StateManager {
  states = {};

  set(name, items) {
    this.states[name] = new StateArray(...items);
    render();
  }

  get(name) {
    return this.states[name];
  }
}

state = new StateManager();


function render() {console.log("rendered", state.states);}


state.set("board", [1,2,3,4]);
state.set("os", [1,2,3,4]);
state.set("snaps", [1,2,3,4]);


state.get("board").remove(3);
state.get("snaps").pop(3);
state.get("os").push(10);
```

Maybe this? If you don't think this works then let's just forget this for now and continue with your code.